api/api/models_api.go

```diff
@@ -100,3 +105,157 @@ func (c *ModelsController) GetModel(r *http.Request, vars map[string]string, bod
+	}
+
+	model, err := c.ModelsService.FindByID(ctx, modelID)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			return NotFound(fmt.Sprintf("Model id %s not found", modelID))
+		}
+		return InternalServerError(err.Error())
+	}
+
+	// get all model version for this model
+	versions, _, err := c.VersionsService.ListVersions(ctx, modelID, c.MonitoringConfig, service.VersionQuery{})
+	if err != nil {
+		return InternalServerError(err.Error())
+	}
+
+	var allJobInModel []*models.PredictionJob
+	var allEndpointInModel []*models.VersionEndpoint
+
+	// iterate for every version, check the active prediction job / active endpoint
+	for _, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// check active prediction job
+			// if there are any active prediction job using the model version, deletion of the model version are prohibited
+			httpStatus, jobs, err := c.checkActivePredictionJobs(ctx, model, version)
+			if err != nil {
+				return NewError(httpStatus, err.Error())
+			}
+
+			// save all prediction jobs for deletion process later on
+			allJobInModel = append(allJobInModel, jobs...)
+
+		} else {
+			// handle for model with type non pyfunc
+			// check active endpoints
+			// if there are any active endpoint using the model version, deletion of the model version are prohibited
+			httpStatus, endpoints, err := c.checkActiveEndpoints(ctx, model, version)
+			if err != nil {
+				return NewError(httpStatus, err.Error())
+			}
+
+			// save all endpoint for deletion process later on
+			allEndpointInModel = append(allEndpointInModel, endpoints...)
+
+		}
+	}
+
+	// DELETING ALL THE RELATED ENTITY
+	for _, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// DELETE PREDICTION JOBS
+			for _, job := range allJobInModel {
+				_, err = c.PredictionJobService.StopPredictionJob(ctx, job.Environment, model, version, job.ID)
+				if err != nil {
+					log.Errorf("failed to stop prediction job %v", err)
+					return InternalServerError(fmt.Sprintf("Failed stopping prediction job: %s", err))
+				}
+			}
+		} else {
+			// DELETE ENDPOINTS
+			for _, endpoint := range allEndpointInModel {
+				err = c.EndpointsService.DeleteEndpoint(version, endpoint)
+				if err != nil {
+					log.Errorf("failed to undeploy endpoint job %v", err)
+					return InternalServerError(fmt.Sprintf("Failed to delete endpoint: %s", err))
+				}
+			}
+		}
+		// DELETE VERSION
+		err = c.VersionsService.Delete(version)
+		if err != nil {
+			log.Errorf("failed to delete model version %v", err)
+			return InternalServerError(fmt.Sprintf("Delete Failed: %s", err.Error()))
+		}
+	}
+	if model.Type != "pyfunc_v2" {
+		// undeploy all existing model endpoint
+		modelEndpoints, err := c.ModelEndpointsService.ListModelEndpoints(ctx, modelID)
+		if err != nil {
+			return InternalServerError(err.Error())
+		}
+
+		for _, modelEndpoint := range modelEndpoints {
+			_, err := c.ModelEndpointsService.UndeployEndpoint(ctx, model, modelEndpoint)
```

Comment:
Hmm it seems like `UndeployEndpoint` will only call `modelEndpointStorage.Save`, which only [updates](https://github.com/caraml-dev/merlin/blob/d0d1ecdda927482b666ca189662c5419b16169a2/api/service/model_endpoint_service.go#L204) the model endpoint in the endpoints database instead of actually deleting it completely, should there instead be a new function for the `ModelEndpointStorage` interface that's actually implemented to `Delete` endpoints? 🤔 Interestingly, the `EndpointsService.DeleteEndpoint` being called above actually performs a `Delete` operation on the version endpoints database.

A slightly unrelated question, but if all the version endpoints have already been deleted (or undeployed rather) by this point, is it still possible that a model endpoint still exists? Or is this section redundant (there is logically no way a model endpoint can still exist if all the version endpoints have been deleted)?

Reply:
oops, sorry for missing this out, yes, it should be DeleteModelEndpoint, since the prerequisite of model deletion is no more active deployment, so we can just immediately delete it from the db. I already update it so it will immediately delete the model endpoint data from the database.

Reply:
Sounds good! Thanks for making that change!